sycl/doc/SYCLPluginInterface.md

```diff
@@ -0,0 +1,318 @@
+the SYCL runtime runs. TBD - describe the process in details.
+
+#### Plugin binary interface
+TBD - list and describe all the symbols plugin must export in order to be picked
+up by the SYCL runtime for offload.
+
+#### OpenCL plugin
+
+OpenCL plugin is a usual plugin from SYCL runtime standpoint, but its loading
+and initialization involves a nested discovery process which finds out available
+OpenCL implementations. They can be installed either in the standard Khronos
+ICD-compatible way or not, and the OpenCL plugin can hook up with both.
+
+##### Linking to ICD-compatible OpenCL installations
+
+The plugin dynamically loads the OCL ICD loader - libOpenCL.so from system
+paths, and on success it asks the loader to provide all OpenCL implementations
+installed on the system. To do that it finds and calls the "clGetPlatformIDs"
+in the ICD loader and records the returned platform objects for later use. The
+plugin may choose to dynamically link with libOpenCL.so at build time.
+ 
+##### Linking to ICD-incompatible OpenCL installations
+
+For the OpenCL plugin to hook up such installation, it must be described in a
+configuration file similarly to plugin configuration file. *TBD - location and
+format of custom OpenCL installation config file*. The plugin will dynamically
+load the dynamic library pointed to in the configuration file and query it for
+OpenCL API function addresses similarly ICD loader. In more details, it will:
+- search for  the ```clGetExtensionFunctionAddress``` function in the 
+  library and call it with the string *�clIcdGetPlatformIDsKHR�* to find
+  the ```clIcdGetPlatformIDsKHR``` function pointer
+- then, using ```clIcdGetPlatformIDsKHR```, find the set of
+  ```cl_platform_id``` objects that the library provides, and store them
+  internally to enumerate available devices when required
+
+### Device enumeration by plugins
+
+TBD
+
+## PI API Specification
+
+PI allows for two levels of API implementation - basic and extended. The basic
+level comprises core APIs needed to perform SYCL offload. The extended level is
+needed only for OpenCL interoperability. 
+
+### The basic PI APIs
+
+This level requires availability of functions representing core functionality,
+such as device memory management, kernel creation and parameter setting,
+enqueuing kernel for execution, etc. There are functions which fully match
+semantics of the corresponding OpenCL X.X functions, as well as functions which
+don't have direct mapping to OpenCL. The two lists are below.
+ 
+*__OpenCL-based APIs__*
+
+    piGetPlatformIDs
+    piGetDeviceIDs
+    piCreateContext
+    piReleaseContext
+    piCreateCommandQueue
+    piReleaseCommandQueue
+    piCreateBuffer
+    piReleaseMemObject
+    piReleaseSampler
+    piReleaseProgram
+    piBuildProgram
+    piCreateKernel
+    piReleaseKernel
+    piSetKernelArg
+    piWaitForEvents
+    piReleaseEvent
+    piGetEventProfilingInfo
+    piFlush
+    piFinish
+    piEnqueueReadBuffer
```

Comment:
I could be wrong or misunderstanding it, but I think this list is missing some of the calls from the pi.h interface (it may be intentional though as its the minimal subset). For example piEnqueueReadBufferRect and piEnqueueWriteBufferRect which are used for handling 2d/3d buffer transfers.

Reply:
Should we maybe avoid this duplication then and maintain APIs in one place - pi.h?

Reply:
I think just directing people to the relevant place to find the plugin interface (pi.h) and giving an example interface call would work, so they know what they're looking for. It means you wouldn't have to maintain the list in the .md every time pi.h changed.

Reply:
Doxygen or similar could be used to generate a more readable documentation directly from the headers

Reply:
> It means you wouldn't have to maintain the list in the .md every time pi.h changed.

Right, that's what I meant by avoiding duplication. Will remove most of the APIs and leave just few examples.

> Doxygen or similar could be used to generate a more readable documentation directly from the headers

Good idea. Should be doxygen, I think, to be compatible with llvm/clang. We can start with maintaining doxygen in the pi.h